linkerd/http-route/src/http/filter/redirect.rs

```diff
@@ -0,0 +1,100 @@
+use super::ModifyPath;
+use crate::http::RouteMatch;
+use http::{
+    uri::{Authority, InvalidUri, Scheme, Uri},
+    StatusCode,
+};
+
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct RedirectRequest {
+    pub scheme: Option<Scheme>,
+    pub host: Option<String>,
+    pub port: Option<u16>,
+    pub path: Option<ModifyPath>,
+    pub status: Option<StatusCode>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidRedirect {
+    #[error("redirects may only replace the path prefix when a path prefix match applied")]
+    InvalidReplacePrefix,
+
+    #[error("redirect produced an invalid location: {0}")]
+    InvalidLocation(#[from] http::Error),
+
+    #[error("redirect produced an invalid authority: {0}")]
+    InvalidAuthority(#[from] InvalidUri),
+
+    #[error("no authority to redirect to")]
+    MissingAuthority,
+}
+
+#[derive(Clone, Debug)]
+pub struct Redirection {
+    pub status: http::StatusCode,
+    pub location: Uri,
+}
+
+// === impl RedirectRequest ===
+
+impl RedirectRequest {
+    pub fn apply(
+        &self,
+        orig_uri: &http::Uri,
+        rm: &RouteMatch,
+    ) -> Result<Option<Redirection>, InvalidRedirect> {
+        let location = {
+            let scheme = self
+                .scheme
+                .clone()
+                .or_else(|| orig_uri.scheme().cloned())
+                .unwrap_or(Scheme::HTTP);
+
+            let authority: Authority = match (self.host.as_deref(), self.port) {
+                // If a host is configured, use it and whatever port is configured.
+                (Some(h), p) => p
+                    .map(|p| format!("{}:{}", h, p).parse())
+                    .unwrap_or_else(|| h.parse())?,
+                // If a host is NOT configured, use the request's original host and either an
+                // overridden port or the original port.
+                (None, p) => {
+                    let h = orig_uri.host().ok_or(InvalidRedirect::MissingAuthority)?;
+                    p.or_else(|| orig_uri.port_u16())
+                        .map(|p| format!("{}:{}", h, p).parse())
+                        .unwrap_or_else(|| h.parse())?
+                }
+            };
+
+            let path = {
+                use crate::http::r#match::PathMatch;
+
+                let orig_path = orig_uri.path();
+                match &self.path {
+                    None => orig_path.to_string(),
+                    Some(ModifyPath::ReplaceFullPath(p)) => p.clone(),
+                    Some(ModifyPath::ReplacePrefixMatch(new_pfx)) => match rm.route.path() {
+                        PathMatch::Prefix(pfx_len) if *pfx_len <= orig_path.len() => {
+                            let (_, rest) = orig_path.split_at(*pfx_len);
+                            format!("{}{}", new_pfx, rest)
+                        }
+                        _ => return Err(InvalidRedirect::InvalidReplacePrefix),
+                    },
+                }
+            };
```

Comment:
Noticed what I believe is a bug here: in all cases (even when just using the original path), this filter will clobber any query params. This is because we get `orig_path` by calling `orig_uri.path()`, [which returns _just_ the path component off the URI](https://docs.rs/http/latest/http/uri/struct.Uri.html#method.path), and not the query component, but then we pass a value constructed using `orig_path` to `uri::Builder::path_and_query` when constructing the new URI. If the original URL had a query component, it's discarded here.

I think we should probably call [`orig_uri.query()`](https://docs.rs/http/latest/http/uri/struct.Uri.html#method.query) at some point and ensure that the query parameters are added to the new path-and-query part passed to `uri::Builder::path_and_query`.

Reply:
(as a side note, it also appears that the redirect filter will currently always throw out any [fragment identifier](https://www.rfc-editor.org/rfc/rfc3986#section-3.5) component of the URI...but unfortunately, i don't think we can actually fix that one in Linkerd, because the `http` crate's `uri` module does not currently appear to expose any APIs for getting or setting a fragment?)